### PR TITLE
fix: prevent funbox reactivation during page transition to leaderboards (@d1rshan)

### PR DESCRIPTION
**Pull request by d1rshan**
### Bug

https://github.com/user-attachments/assets/763fb1ce-8f49-4f74-9695-c77adb34faab

### Cause

When navigating from the test page to leaderboards, the leaderboard's sync loading delays `setActivePage` from firing. During this window, `init()` checks `getActivePage() === "test"` which still returns `"test"` since `setActivePage` hasn't fired yet, causing funboxes to be incorrectly reactivated.

### Fix

Added a new `isTestPageVisible` signal that is set to `false` immediately after the test page finishes hiding (before sync loading starts) and back to `true` after sync loading completes when navigating to test. `init()` now checks `isTestPageVisible()` instead of `getActivePage() === "test"` to determine whether to activate funboxes.

### CODE REVIEW

**Comment by Leonabcd123** [comment]
This actually looks really cool lol, maybe we can add a setting to make funboxes apply globally.

**Comment by d1rshan** [comment]
yea that would be cool for some effects,  i don't know if you watched the vid till the end coz it wouldn't be a good idea for all the funboxes.

**Comment by Leonabcd123** [comment]
> yea that would be cool for some effects, i don't know if you watched the vid till the end coz it wouldn't be a good idea for all the funboxes.

The user could always turn off that setting if they don't want funboxes to apply to the rest of the website. Maybe even have a list of all the funboxes the user wants to have globally applied.

**Comment by d1rshan** [comment]
> The user could always turn off that setting if they don't want funboxes to apply to the rest of the website. Maybe even have a list the user can modify of all the funboxes the user wants to apply globally.


yess that sounds like a good idea :)

i'll see what i can do for "global funboxes" in a new pr coz that doesn't come under the scope of this bug fix,

i'm thinking we can just keep a single toggle control like you said (but without the specific list control for now): `apply funboxes globally - yes/no` (default will be no)  and all the valid funboxes will be applied globally based on this

what do you think?

**Comment by Leonabcd123** [comment]
> i'll see what i can do for "global funboxes" in a new pr coz that doesn't come under the scope of this bug fix,

Let's wait for feedback from miodec or fehmer before starting any work. I'll open a discussion for that (#7874).

> i'm thinking we can just keep a single toggle control like you said (but without the specific list control for now): `apply funboxes globally - yes/no` (default will be no) and all the valid funboxes will be applied globally
> 
> what do you think?

Not sure what you mean by "valid" funboxes, I think that global switch should apply to every funbox (although some won't have a visible effect because they're word generation related). Otherwise sounds good!

**Comment by d1rshan** [comment]
> Let's wait for feedback from miodec or fehmer before starting any work. I'll open a discussion for that.

yeah ofc 👍

> Not sure what you mean by "valid" funboxes, I think that global switch should apply to every funbox (although some won't have a visible effect because they're word generation related). Otherwise sounds good!

by valid, i mean funboxes that visibly affect the ui without breaking it

**Comment by fehmer** [comment]
I am pretty sure the funboxes were cleared when leaving the test page at some point. Maybe you can bisect the change which broke this.

I don't think we should add a setting to activate some funboxes on all pages or just test page, very niche use-case.

For the CRT funbox we added special handling to apply on all pages.

**Comment by d1rshan** [comment]
> I am pretty sure the funboxes were cleared when leaving the test page at some point. Maybe you can bisect the change which broke this.

Yes the funboxes are still being cleared, the bug is an **edge case** where they get reapplied.

**NOTE**:
- this is repro only when we move away from the test page.
- only funbox effects that affect global css elements are noticable, eg: `mirror`, `space_balls` etc and others wouldn't be coz they target selectors that exist in test page only.

---

### Issue

The issue is that when leaving the test page, `afterHide` was still calling `TestLogic.restart({ noAnim: true })`, and `restart()` eventually calls `init()` again. 

 In this edge case, that allowed `Funbox.activate()` to run during the route transition and reapply the active funbox CSS right after it had been cleared.

  That is why it did not happen on every navigation, but could show up after moving between routes multiple times.

---

### Fix
The fix was to keep the existing test page reset, but make that specific teardown restart skip `init()`, so leaving the test page stays cleanup-only and cannot reactivate funboxes during navigation.

**Comment by d1rshan** [comment]
> I don't think we should add a setting to activate some funboxes on all pages or just test page, very niche use-case.

okay, understood.

> For the CRT funbox we added special handling to apply on all pages.

yes, CRT funbox is intentional and it is handling it properly to be applied globally and this PR doesn't affect existing CRT behavior

**Comment by fehmer** [comment]
This is the commit adding this bug:

https://github.com/monkeytypegame/monkeytype/commit/765ca9569feee597d0d42c7ef809349e460fdaad

**Comment by d1rshan** [comment]
> This is the commit adding this bug:
> 
> [765ca95](https://github.com/monkeytypegame/monkeytype/commit/765ca9569feee597d0d42c7ef809349e460fdaad)

Thanks, i was kind of mislead by assuming the bug always existed but went unnoticeable. it looks like 765ca95 exposed the bug.

i found the same timing pattern on:
  - `frontend/src/ts/pages/account.ts` 
  - `frontend/src/ts/pages/friends.ts` 
  
just to confirm, do you want me to address the fix close to the regression point or keep my current fix which addresses the teardown directly?

**Comment by fehmer** [comment]
Can you reproduce the same error when switching from test to account or friends page with funbox mirror active? For me only the switch to the leaderboard does not clear the funbox css.

I think it would be good to find the reason first why this is only broken for the transition to the leaderboard page, than decide how to fix it.

**Comment by d1rshan** [comment]
ohh yes, weirdly the css is leaking only for leaderboard and other pages are getting the leaked css only after navigating from leaderboard page (this is what happened at the end of the vid)

i'll manually look into it and will try to figure out the exact reason

**Comment by d1rshan** [comment]
@fehmer, i figured out the exact issue,

so the flow now is:

1. first leaving the test page triggers test restart

2.  the restart method eventually reaches init() and this condition is checked:
https://github.com/monkeytypegame/monkeytype/blob/852d15837edc6e5241d8649828f9b13227817202/frontend/src/ts/test/test-logic.ts#L426-L428

3. _(parallely this happens when navigating to leaderboard page)_
leaderboard page follows synchronous loading ie its promise must resolve before `setActivePage(nextPage.id)` is triggered 
https://github.com/monkeytypegame/monkeytype/blob/852d15837edc6e5241d8649828f9b13227817202/frontend/src/ts/controllers/page-controller.ts#L57-L65


4.  `getActivePage()` inside `init()` still returns `"test"` because the active page has not been updated to `"leaderboards"` yet. so `init()` assumes the test page is still active and reactivates the funboxes.

and also since leaderboard doesn't bother about cleaning "funbox css" after we move to other pages from leaderboard, the css remains and other pages get affected.

File: frontend/src/ts/test/test-logic.ts [hunk 1]
```diff
@@ -423,7 +423,7 @@ async function init(): Promise<boolean> {
     return await init();
   }
 
-  if (getActivePage() === "test") {
+  if (getActivePage() === "test" && !PageTransition.get()) {
     await Funbox.activate();
```
[the remark this answers is not in the document]
Comment (d1rshan) [reply frontend/src/ts/test/test-logic.ts]: irrelevant cause accidentally requested review with older pr description fix

File: frontend/src/ts/test/test-logic.ts [hunk 2]
```diff
@@ -423,7 +423,7 @@ async function init(): Promise<boolean> {
     return await init();
   }
 
-  if (getActivePage() === "test") {
+  if (getActivePage() === "test" && !PageTransition.get()) {
```
[the remark this answers is not in the document]
Comment (d1rshan) [reply frontend/src/ts/test/test-logic.ts]: forgot to push other files 😭

**Comment by d1rshan** [comment]
hey @fehmer, i updated the fix to a cleaner approach. let me know if you'd like any changes

> ### Fix
> 
> Added a new `isTestPageVisible` signal that is set to `false` immediately after the test page finishes hiding (before sync loading starts) and back to `true` after sync loading completes when navigating to test. `init()` now checks `isTestPageVisible()` instead of `getActivePage() === "test"` to determine whether to activate funboxes.

**Comment by d1rshan** [comment]
Closing cause not relevant anymore
